src/client/content_cao.cpp

```diff
@@ -1817,10 +1860,51 @@ void GenericCAO::processMessage(const std::string &data)
 		updateAnimationSpeed();
 	} else if (cmd == AO_CMD_SET_BONE_POSITION) {
 		std::string bone = deSerializeString16(is);
-		v3f position = readV3F32(is);
-		v3f rotation = readV3F32(is);
-		m_bone_position[bone] = core::vector2d<v3f>(position, rotation);
-
+		const BoneOverride *previous = m_bone_override[bone];
+		BoneOverride *props = new BoneOverride();
+		bool is_identity = false; // whether this is an identity override to be removed
+		props->position.vector = readV3F32(is);
+		props->rotation.next = core::quaternion(readV3F32(is) * core::DEGTORAD);
+		props->scale.vector = readV3F32(is); // reads past end of string on older cmds
+		if (is.eof()) {
+			// Backwards compatibility for protocol versions <= 40
+			props->scale.vector = v3f(1, 1, 1); // restore the scale which was not sent
+			props->position.absolute = true;
+			props->rotation.absolute = true;
+		} else {
+			props->position.interpolation_duration = readF32(is);
+			props->rotation.interpolation_duration = readF32(is);
+			props->scale.interpolation_duration = readF32(is);
+			u8 absoluteFlag = readU8(is);
+			if (absoluteFlag == 0
+					&& props->position.vector == v3f(0.0f, 0.0f, 0.0f)
+					&& props->rotation.next == core::quaternion()
+					&& props->scale.vector == v3f(1.0f, 1.0f, 1.0f)) {
+				is_identity = true;
+			} else {
+				props->position.absolute = (absoluteFlag & 1) > 0;
+				props->rotation.absolute = (absoluteFlag & 2) > 0;
+				props->scale.absolute = (absoluteFlag & 4) > 0;
+				if (previous) {
+					props->position.previous = previous->position.vector;
+					props->rotation.previous = previous->rotation.next;
+					props->scale.previous = previous->scale.vector;
+				} else {
+					// disable interpolation
+					props->position.interpolation_duration = 0.0f;
+					props->rotation.interpolation_duration = 0.0f;
+					props->scale.interpolation_duration = 0.0f;
+				}
+			}
+		}
+		if (previous)
```

Comment:
it doesn't look necessary here to create a new object instead of reusing whats already there